master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,470 @@
+	s3client := s3.New(sess)
+
+	var errs []error
+	downloader := s3manager.NewDownloader(sess, func(d *s3manager.Downloader) {
+		d.Concurrency = 1 // Setting concurrency to 1 to use seqWriterAt
+	})
+	funcReadPage := func(output *s3.ListObjectsV2Output, lastPage bool) bool {
+		iter := newBatchDownloadIterator(d.aw, d.bucket, output.Contents)
+		// Download every bucket in this page
+		err = downloader.DownloadWithIterator(c, iter)
+		if iter.Err() != nil {
+			errs = append(errs, iter.Err())
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		// Return False to stop paging
+		return len(errs) == 0
+	}
+	err = s3client.ListObjectsV2PagesWithContext(
+		c,
+		&s3.ListObjectsV2Input{
+			Bucket: &d.bucket,
+			Prefix: &d.prefix,
+		},
+		funcReadPage,
+	)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		msg := "one or more errors encountered during checkpoint download:"
+		for _, v := range errs {
+			msg += fmt.Sprintf("\n  %s;", v.Error())
+		}
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func newDownloader(
+	storageConfig *expconf.CheckpointStorageConfig,
+	aw archiveWriter,
+	id string,
+) (checkpointDownloader, error) {
+	switch storage := storageConfig.GetUnionMember().(type) {
+	case expconf.S3Config:
+		return &s3Downloader{
+			aw:     aw,
+			bucket: storage.Bucket(),
+			prefix: strings.TrimLeft(*storage.Prefix()+"/"+id, "/"),
+		}, nil
+	default:
+		return nil, echo.NewHTTPError(http.StatusNotImplemented,
+			fmt.Sprintf("checkpoint download via master is only supported on S3"+
+				", but the checkpoint's storage type is %s", storageConfig2Str(storage)))
+	}
+}
+
+// It is assumed that a http status code is not sent until the first write to w.
+func buildWriterPipeline(w io.Writer, mimeType string) (archiveWriter, error) {
+	// DelayWriter delays the first write until we have successfully downloaded
+	// some bytes and are more confident that the download will succeed.
+	dw := newDelayWriter(w, 16*1024)
+	closers := []io.Closer{dw}
+	switch mimeType {
+	case MIMEApplicationGZip:
+		gz := gzip.NewWriter(dw)
+		closers = append(closers, gz)
+
+		tw := tar.NewWriter(gz)
+		closers = append(closers, tw)
+
+		return &tarArchiveWriter{archiveClosers{closers}, tw}, nil
+
+	case MIMEApplicationZip:
+		zw := zip.NewWriter(dw)
+		closers = append(closers, zw)
+
+		return &zipArchiveWriter{archiveClosers{closers}, zw, nil}, nil
+
+	default:
+		return nil, fmt.Errorf(
+			"MIME type must be %s or %s but got %s",
+			MIMEApplicationGZip, MIMEApplicationZip, mimeType)
+	}
+}
+
+func (m *Master) getCheckpointStorageConfig(id uuid.UUID) (
+	*expconf.CheckpointStorageConfig, error) {
+	checkpoint, err := m.db.CheckpointByUUID(id)
+	if err != nil || checkpoint == nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(checkpoint.CheckpointTrainingMetadata.ExperimentConfig)
+	if err != nil {
+		return nil, err
+	}
+	spew.Dump(bytes)
+	spew.Dump(checkpoint.CheckpointTrainingMetadata.ExperimentConfig)
```

Comment:
is this leftover from some debugging?

Reply:
Yeah, I though I deleted all of them. Let me remove them.